src/control/lib/hardware/pci_test.go

```diff
@@ -0,0 +1,534 @@
+	case 2:
+		args = append(args, args[1]+1)
+	}
+
+	return &PCIBus{
+		LowAddress:  *MockPCIAddress(args[0], args[1]),
+		HighAddress: *MockPCIAddress(args[0], args[2]),
+	}
+}
+
+func mockPCIDevice(name string, addr ...uint8) *PCIDevice {
+	if len(addr) == 0 {
+		addr = []uint8{0}
+	}
+
+	return &PCIDevice{
+		Name:    fmt.Sprintf("%s%02d", name, addr[0]),
+		PCIAddr: *MockPCIAddress(addr...),
+	}
+}
+
+func (d *PCIDevice) withType(t DeviceType) *PCIDevice {
+	d.Type = t
+	return d
+}
+
+func (d *PCIDevice) withLinkSpeed(ls float64) *PCIDevice {
+	d.LinkSpeed = ls
+	return d
+}
+
+func TestHardware_NewPCIAddress(t *testing.T) {
+	for name, tc := range map[string]struct {
+		addrStr string
+		expVMD  *PCIAddress
+		expErr  error
+	}{
+		"valid": {
+			addrStr: "0000:80:00.0",
+		},
+		"invalid": {
+			addrStr: "0000:gg:00.0",
+			expErr:  errors.New("parsing \"gg\""),
+		},
+		"invalid vmd": {
+			addrStr: "1ffff:01:00.0",
+			expErr:  errors.New("unexpected length of vmd"),
+		},
+		"vmd address": {
+			addrStr: "0000:5d:05.5",
+		},
+		"vmd backing device address": {
+			addrStr: "ffffff:01:00.0",
+			expVMD: &PCIAddress{
+				Bus:      0xff,
+				Device:   0xff,
+				Function: 0xff,
+			},
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			addr, err := NewPCIAddress(tc.addrStr)
+			common.CmpErr(t, tc.expErr, err)
+			if tc.expErr != nil {
+				return
+			}
+
+			if diff := cmp.Diff(tc.addrStr, addr.String()); diff != "" {
+				t.Fatalf("unexpected string (-want, +got):\n%s\n", diff)
+			}
+			if diff := cmp.Diff(tc.expVMD, addr.VMDAddr); diff != "" {
+				t.Fatalf("unexpected VMD address (-want, +got):\n%s\n", diff)
+			}
+		})
+	}
+}
+
+func TestHardware_PCIAddress_Equals(t *testing.T) {
+	for name, tc := range map[string]struct {
+		addrStr1 string
+		addrStr2 string
+		expEqual bool
+	}{
+		"same": {
+			addrStr1: "0000:80:00.0",
+			addrStr2: "0000:80:00.0",
+			expEqual: true,
+		},
+		"different domain": {
+			addrStr1: "0000:80:00.0",
+			addrStr2: "0001:80:00.0",
+			expEqual: false,
+		},
+		"different bus": {
+			addrStr1: "0000:80:00.0",
+			addrStr2: "0000:81:00.0",
+			expEqual: false,
+		},
+		"different device": {
+			addrStr1: "0000:80:00.0",
+			addrStr2: "0000:80:01.0",
+			expEqual: false,
+		},
+		"different function": {
+			addrStr1: "0000:80:00.0",
+			addrStr2: "0000:80:00.1",
+			expEqual: false,
+		},
+		"vmd backing device same": {
+			addrStr1: "ffffff:01:00.0",
+			addrStr2: "ffffff:01:00.0",
+			expEqual: true,
+		},
+		"vmd backing device different": {
+			addrStr1: "ffffff:01:00.0",
+			addrStr2: "fffffe:01:00.0",
+			expEqual: false,
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			addr1, err := NewPCIAddress(tc.addrStr1)
+			if err != nil {
+				t.Fatal(err)
+			}
+			addr2, err := NewPCIAddress(tc.addrStr2)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			if diff := cmp.Diff(tc.expEqual, addr1.Equals(addr2)); diff != "" {
+				t.Fatalf("unexpected result (-want, +got):\n%s\n", diff)
+			}
```

Comment:
cmp.Diff feels a bit like overkill comparing two bools

Reply:
Yeah, probably. I prefer the output format though. ;)